File: backend/app/api/v1/okrs.py
"""OKR API。"""
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import PermissionChecker
from app.database import get_db
from app.models.user import User
from app.schemas.okr import (
    KeyResultCreate,
    KeyResultOut,
    KeyResultUpdate,
    OkrCreate,
    OkrDetailOut,
    OkrListOut,
    OkrOut,
    OkrStatsOut,
    OkrUpdate,
)
from app.services import okr as okr_service

router = APIRouter(prefix="/okrs", tags=["OKR"])


@router.get("/stats", response_model=OkrStatsOut, summary="OKR 统计")
def stats(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
    period_label: Optional[str] = None,
) -> OkrStatsOut:
    return OkrStatsOut(**okr_service.okr_stats(db, current_user, period_label=period_label))


@router.get("", response_model=OkrListOut, summary="OKR 列表")
def list_okrs(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
    status: Optional[str] = None,
    level: Optional[str] = None,
    period_label: Optional[str] = None,
    keyword: Optional[str] = None,
    scope: Optional[str] = Query(None, description="mine/all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> OkrListOut:
    total, items = okr_service.list_okrs(
        db,
        current_user,
        status=status,
        level=level,
        period_label=period_label,
        keyword=keyword,
        scope_filter=scope,
        page=page,
        page_size=page_size,
    )
    return OkrListOut(total=total, items=[OkrOut.model_validate(x) for x in items])


@router.post("", response_model=OkrOut, summary="创建 OKR")
def create_okr(
    payload: OkrCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> OkrOut:
    return OkrOut.model_validate(okr_service.create_okr(db, current_user, payload))


@router.get("/{okr_id}", response_model=OkrDetailOut, summary="OKR 详情")
def get_okr(
    okr_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> OkrDetailOut:
    return OkrDetailOut.model_validate(okr_service.get_okr_detail(db, current_user, okr_id))


@router.patch("/{okr_id}", response_model=OkrOut, summary="编辑 OKR")
def update_okr(
    okr_id: int,
    payload: OkrUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> OkrOut:
    return OkrOut.model_validate(okr_service.update_okr(db, current_user, okr_id, payload))


@router.post("/{okr_id}/confirm", response_model=OkrOut, summary="确认目标")
def confirm_okr(
    okr_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> OkrOut:
    return OkrOut.model_validate(okr_service.confirm_okr(db, current_user, okr_id))


@router.post("/{okr_id}/complete", response_model=OkrOut, summary="完成目标")
def complete_okr(
    okr_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> OkrOut:
    return OkrOut.model_validate(okr_service.complete_okr(db, current_user, okr_id))


@router.post("/{okr_id}/terminate", response_model=OkrOut, summary="终止目标")
def terminate_okr(
    okr_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
    reason: Optional[str] = None,
) -> OkrOut:
    return OkrOut.model_validate(okr_service.terminate_okr(db, current_user, okr_id, reason))


@router.post("/{okr_id}/key-results", response_model=KeyResultOut, summary="添加关键结果")
def add_key_result(
    okr_id: int,
    payload: KeyResultCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> KeyResultOut:
    return KeyResultOut.model_validate(okr_service.add_key_result(db, current_user, okr_id, payload))


@router.patch(
    "/{okr_id}/key-results/{kr_id}",
    response_model=KeyResultOut,
    summary="更新关键结果/进度",
)
def update_key_result(
    okr_id: int,
    kr_id: int,
    payload: KeyResultUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["okr:view"]))],
) -> KeyResultOut:
    return KeyResultOut.model_validate(
        okr_service.update_key_result(db, current_user, okr_id, kr_id, payload)
    )

File: frontend/src/api/okrs.ts
/**
 * OKR API
 */
import request from './request'

export type OkrStatus = 'pending' | 'active' | 'completed' | 'adjusted' | 'terminated'
export type OkrLevel = 'company' | 'department' | 'personal'

export interface KeyResult {
  id: number
  okr_id: number
  title: string
  target_value: number | string
  current_value: number | string
  unit?: string | null
  weight: number
  sort_order: number
  remark?: string | null
  created_at: string
  updated_at: string
  progress?: number
}

export interface Okr {
  id: number
  title: string
  level: OkrLevel | string
  period_type: string
  period_label: string
  status: OkrStatus | string
  owner_id?: number | null
  creator_id?: number | null
  department_id?: number | null
  parent_id?: number | null
  progress: number
  description?: string | null
  remark?: string | null
  created_at: string
  updated_at: string
  owner_name?: string | null
  creator_name?: string | null
  parent_title?: string | null
  kr_count?: number
}

export interface OkrDetail extends Okr {
  key_results: KeyResult[]
}

export interface OkrStats {
  total: number
  pending: number
  active: number
  completed: number
  terminated: number
  mine: number
  avg_progress?: number
  unaligned?: number
  company_count?: number
  department_count?: number
  personal_count?: number
  risk_count?: number
}

export const OKR_STATUS_LABEL: Record<string, string> = {
  pending: '待确认',
  active: '进行中',
  completed: '已完成',
  adjusted: '已调整',
  terminated: '已终止',
}

export const OKR_LEVEL_OPTIONS = [
  { value: 'company', label: '公司级' },
  { value: 'department', label: '部门级' },
  { value: 'personal', label: '个人级' },
]

export const OKR_PERIOD_OPTIONS = [
  { value: 'yearly', label: '年度' },
  { value: 'quarterly', label: '季度' },
  { value: 'monthly', label: '月度' },
]

export function fetchOkrStats(period_label?: string) {
  return request.get<OkrStats>('/okrs/stats', {
    params: period_label ? { period_label } : undefined,
  })
}

export function fetchOkrs(params: {
  status?: string
  level?: string
  period_label?: string
  keyword?: string
  scope?: string
  page?: number
  page_size?: number
}) {
  return request.get<{ total: number; items: Okr[] }>('/okrs', { params })
}

export function fetchOkrDetail(id: number) {
  return request.get<OkrDetail>(`/okrs/${id}`)
}

export function createOkr(data: {
  title: string
  level?: string
  period_type?: string
  period_label: string
  description?: string
  parent_id?: number
  remark?: string
  key_results?: Array<{
    title: string
    target_value?: number
    current_value?: number
    unit?: string
    weight?: number
  }>
}) {
  return request.post<Okr>('/okrs', data)
}

export function updateOkr(id: number, data: Partial<Okr>) {
  return request.patch<Okr>(`/okrs/${id}`, data)
}

export function confirmOkr(id: number) {
  return request.post<Okr>(`/okrs/${id}/confirm`)
}

export function completeOkr(id: number) {
  return request.post<Okr>(`/okrs/${id}/complete`)
}

export function terminateOkr(id: number, reason?: string) {
  return request.post<Okr>(`/okrs/${id}/terminate`, null, { params: { reason } })
}

export function addKeyResult(
  okrId: number,
  data: {
    title: string
    target_value?: number
    current_value?: number
    unit?: string
    weight?: number
  },
) {
  return request.post<KeyResult>(`/okrs/${okrId}/key-results`, data)
}

export function updateKeyResult(okrId: number, krId: number, data: Partial<KeyResult>) {
  return request.patch<KeyResult>(`/okrs/${okrId}/key-results/${krId}`, data)
}
